### Weekly assignment 4, secondary school students Imke Schrijen

For this weekly assignment, I will be predicting if students are romantic using other variables in the dataset. I will use Random Forest, this is a machine learning algorithm that consists of different decision trees. This number of decision trees is determined by yourself and depends on the date and contextual factors. Each decision tree contains a different part of the data, which prevents outliers from having a (negative) influence on the outcome of the prediction.

In [51]:
import sklearn as sk
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

### Importing the data

First, I will have a look at the data, then I will create a new dataset in which only the columns I use for the prediction are included

In [54]:
df = pd.read_csv('student-por.csv')
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0  11  11  
1      5        3      3     1     1      3        2   9  11  11  
2      4        3      2     2     3      3        6  12  13  12  
3      3        2      2     1     1      5        0  14  14  14  
4      4        3      2     1     2      5        0  11  13  13  

[5 rows x 33 columns]

In [56]:
df_new = df[['famrel', 'age', 'freetime', 'studytime', 'goout', 'traveltime', 'absences', 'romantic']]

I chose these variables because all of them are numeric. In addition, I think that people who are in a relationship may be more often absent, have a lot of free time after school (to do something with someone) and go out often. I wanted to include the column Sex as well, but I did not succeed in converting the F and M to 0 and 1. I hope this will give me a good outcome aswell.

In [57]:
df_new.head()

famrel  age  freetime  studytime  goout  traveltime  absences romantic
0       4   18         3          2      4           2         4       no
1       5   17         3          2      3           1         2       no
2       4   15         3          2      2           1         6       no
3       3   15         2          3      2           1         0      yes
4       4   16         3          2      2           1         0       no

In the code above I've counted how much students are and how much students are not in a relationship. This means that 36,8% of the students are in a relationship. Now we will get the x and y for the predictions.

In [58]:
X = df.loc[:,'famrel':'absences'] #This gets all the columns from Sex to Absences
y = df['romantic']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

### Training the algorithm

In [71]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=1, n_estimators=100) #RF is a random algorithm, so to get the same results we need to use random_state
rf = rf.fit(X_train, y_train)

In [72]:
rf.score(X_test,y_test)

0.5897435897435898

### Evaluating the model

In [73]:
y_pred = rf.predict(X_test) #the predicted values
cm = confusion_matrix(y_test, y_pred) #creates a "confusion matrix"
cm = pd.DataFrame(cm, index=['Not in a relationship (actual)', 'In a relationship (actual)'], columns = ['Not in a relationship (pred)', 'In a relationship (pred)']) #label and make df
cm

Not in a relationship (pred)  \
Not in a relationship (actual)                           101   
In a relationship (actual)                                53   

                                In a relationship (pred)  
Not in a relationship (actual)                        27  
In a relationship (actual)                            14

In [74]:
14/41

0.34146341463414637

In [75]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          no       0.66      0.79      0.72       128
         yes       0.34      0.21      0.26        67

    accuracy                           0.59       195
   macro avg       0.50      0.50      0.49       195
weighted avg       0.55      0.59      0.56       195



The precision of the prediction is not very good, about 34% of the students turn out not to have a relationship although it was predicted.

The recall is also very low, the algorithm missed 79% of the students who have a relationship. 

I liked this weekly assignment because it is preparing me for the final assignment. I haven't put a lot of markdowns in between the codes, because I am still struggeling a bit with knowing what everything means. 